File: typescript/src/program/accounts.ts
import { PublicKey } from "@solana/web3.js";
import { DISCRIMINATOR, ACCOUNT_SIZE } from "./constants";
import {
  Exchange,
  Market,
  MarketStatus,
  OrderStatus,
  UserNonce,
  Position,
  Orderbook,
} from "./types";
import { fromLeBytes } from "./utils";

// ============================================================================
// DISCRIMINATOR VALIDATION
// ============================================================================

/**
 * Validate that a buffer starts with the expected discriminator
 */
function validateDiscriminator(
  data: Buffer,
  expected: Buffer,
  accountType: string
): void {
  const actual = data.subarray(0, 8);
  if (!actual.equals(expected)) {
    throw new Error(
      `Invalid ${accountType} discriminator. Expected ${expected.toString("hex")}, got ${actual.toString("hex")}`
    );
  }
}

/**
 * Check if a buffer has a valid Exchange discriminator
 */
export function isExchangeAccount(data: Buffer): boolean {
  if (data.length < 8) return false;
  return data.subarray(0, 8).equals(DISCRIMINATOR.EXCHANGE);
}

/**
 * Check if a buffer has a valid Market discriminator
 */
export function isMarketAccount(data: Buffer): boolean {
  if (data.length < 8) return false;
  return data.subarray(0, 8).equals(DISCRIMINATOR.MARKET);
}

/**
 * Check if a buffer has a valid OrderStatus discriminator
 */
export function isOrderStatusAccount(data: Buffer): boolean {
  if (data.length < 8) return false;
  return data.subarray(0, 8).equals(DISCRIMINATOR.ORDER_STATUS);
}

/**
 * Check if a buffer has a valid UserNonce discriminator
 */
export function isUserNonceAccount(data: Buffer): boolean {
  if (data.length < 8) return false;
  return data.subarray(0, 8).equals(DISCRIMINATOR.USER_NONCE);
}

/**
 * Check if a buffer has a valid Position discriminator
 */
export function isPositionAccount(data: Buffer): boolean {
  if (data.length < 8) return false;
  return data.subarray(0, 8).equals(DISCRIMINATOR.POSITION);
}

/**
 * Check if a buffer has a valid Orderbook discriminator
 */
export function isOrderbookAccount(data: Buffer): boolean {
  if (data.length < 8) return false;
  return data.subarray(0, 8).equals(DISCRIMINATOR.ORDERBOOK);
}

// ============================================================================
// ACCOUNT DESERIALIZATION
// ============================================================================

/**
 * Deserialize Exchange account data
 *
 * Layout (88 bytes):
 * - discriminator: [u8; 8]
 * - authority: Pubkey (32 bytes)
 * - operator: Pubkey (32 bytes)
 * - market_count: u64 (8 bytes)
 * - paused: u8 (1 byte)
 * - bump: u8 (1 byte)
 * - _padding: [u8; 6]
 */
export function deserializeExchange(data: Buffer): Exchange {
  if (data.length < ACCOUNT_SIZE.EXCHANGE) {
    throw new Error(
      `Invalid Exchange data length: ${data.length}, expected ${ACCOUNT_SIZE.EXCHANGE}`
    );
  }

  validateDiscriminator(data, DISCRIMINATOR.EXCHANGE, "Exchange");

  let offset = 0;

  const discriminator = data.subarray(offset, offset + 8);
  offset += 8;

  const authority = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const operator = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const marketCount = fromLeBytes(data.subarray(offset, offset + 8));
  offset += 8;

  const paused = data[offset] !== 0;
  offset += 1;

  const bump = data[offset];
  offset += 1;

  // Skip padding: 6 bytes

  return {
    discriminator,
    authority,
    operator,
    marketCount,
    paused,
    bump,
  };
}

/**
 * Deserialize Market account data
 *
 * Layout (120 bytes):
 * - discriminator: [u8; 8]
 * - market_id: u64 (8 bytes)
 * - num_outcomes: u8 (1 byte)
 * - status: u8 (1 byte)
 * - winning_outcome: u8 (1 byte)
 * - has_winning_outcome: u8 (1 byte)
 * - bump: u8 (1 byte)
 * - _padding: [u8; 3]
 * - oracle: Pubkey (32 bytes)
 * - question_id: [u8; 32]
 * - condition_id: [u8; 32]
 */
export function deserializeMarket(data: Buffer): Market {
  if (data.length < ACCOUNT_SIZE.MARKET) {
    throw new Error(
      `Invalid Market data length: ${data.length}, expected ${ACCOUNT_SIZE.MARKET}`
    );
  }

  validateDiscriminator(data, DISCRIMINATOR.MARKET, "Market");

  let offset = 0;

  const discriminator = data.subarray(offset, offset + 8);
  offset += 8;

  const marketId = fromLeBytes(data.subarray(offset, offset + 8));
  offset += 8;

  const numOutcomes = data[offset];
  offset += 1;

  const statusByte = data[offset];
  offset += 1;

  const winningOutcome = data[offset];
  offset += 1;

  const hasWinningOutcome = data[offset] !== 0;
  offset += 1;

  const bump = data[offset];
  offset += 1;

  // Skip padding: 3 bytes
  offset += 3;

  const oracle = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const questionId = Buffer.from(data.subarray(offset, offset + 32));
  offset += 32;

  const conditionId = Buffer.from(data.subarray(offset, offset + 32));
  offset += 32;

  // Map status byte to enum
  let status: MarketStatus;
  switch (statusByte) {
    case 0:
      status = MarketStatus.Pending;
      break;
    case 1:
      status = MarketStatus.Active;
      break;
    case 2:
      status = MarketStatus.Resolved;
      break;
    case 3:
      status = MarketStatus.Cancelled;
      break;
    default:
      throw new Error(`Unknown market status: ${statusByte}`);
  }

  return {
    discriminator,
    marketId,
    numOutcomes,
    status,
    winningOutcome,
    hasWinningOutcome,
    bump,
    oracle,
    questionId,
    conditionId,
  };
}

/**
 * Deserialize OrderStatus account data
 *
 * Layout (24 bytes):
 * - discriminator: [u8; 8]
 * - remaining: u64 (8 bytes)
 * - is_cancelled: u8 (1 byte)
 * - _padding: [u8; 7]
 */
export function deserializeOrderStatus(data: Buffer): OrderStatus {
  if (data.length < ACCOUNT_SIZE.ORDER_STATUS) {
    throw new Error(
      `Invalid OrderStatus data length: ${data.length}, expected ${ACCOUNT_SIZE.ORDER_STATUS}`
    );
  }

  validateDiscriminator(data, DISCRIMINATOR.ORDER_STATUS, "OrderStatus");

  let offset = 0;

  const discriminator = data.subarray(offset, offset + 8);
  offset += 8;

  const remaining = fromLeBytes(data.subarray(offset, offset + 8));
  offset += 8;

  const isCancelled = data[offset] !== 0;
  offset += 1;

  // Skip padding: 7 bytes

  return {
    discriminator,
    remaining,
    isCancelled,
  };
}

/**
 * Deserialize UserNonce account data
 *
 * Layout (16 bytes):
 * - discriminator: [u8; 8]
 * - nonce: u64 (8 bytes)
 */
export function deserializeUserNonce(data: Buffer): UserNonce {
  if (data.length < ACCOUNT_SIZE.USER_NONCE) {
    throw new Error(
      `Invalid UserNonce data length: ${data.length}, expected ${ACCOUNT_SIZE.USER_NONCE}`
    );
  }

  validateDiscriminator(data, DISCRIMINATOR.USER_NONCE, "UserNonce");

  let offset = 0;

  const discriminator = data.subarray(offset, offset + 8);
  offset += 8;

  const nonce = fromLeBytes(data.subarray(offset, offset + 8));
  offset += 8;

  return {
    discriminator,
    nonce,
  };
}

/**
 * Deserialize Position account data
 *
 * Layout (80 bytes):
 * - discriminator: [u8; 8]
 * - owner: Pubkey (32 bytes)
 * - market: Pubkey (32 bytes)
 * - bump: u8 (1 byte)
 * - _padding: [u8; 7]
 */
export function deserializePosition(data: Buffer): Position {
  if (data.length < ACCOUNT_SIZE.POSITION) {
    throw new Error(
      `Invalid Position data length: ${data.length}, expected ${ACCOUNT_SIZE.POSITION}`
    );
  }

  validateDiscriminator(data, DISCRIMINATOR.POSITION, "Position");

  let offset = 0;

  const discriminator = data.subarray(offset, offset + 8);
  offset += 8;

  const owner = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const market = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const bump = data[offset];
  offset += 1;

  // Skip padding: 7 bytes

  return {
    discriminator,
    owner,
    market,
    bump,
  };
}

/**
 * Deserialize Orderbook account data
 *
 * Layout (144 bytes):
 * - discriminator: [u8; 8]
 * - market: Pubkey (32 bytes)
 * - mint_a: Pubkey (32 bytes)
 * - mint_b: Pubkey (32 bytes)
 * - lookup_table: Pubkey (32 bytes)
 * - bump: u8 (1 byte)
 * - _padding: [u8; 7]
 */
export function deserializeOrderbook(data: Buffer): Orderbook {
  if (data.length < ACCOUNT_SIZE.ORDERBOOK) {
    throw new Error(
      `Invalid Orderbook data length: ${data.length}, expected ${ACCOUNT_SIZE.ORDERBOOK}`
    );
  }

  validateDiscriminator(data, DISCRIMINATOR.ORDERBOOK, "Orderbook");

  let offset = 0;

  const discriminator = data.subarray(offset, offset + 8);
  offset += 8;

  const market = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const mintA = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const mintB = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const lookupTable = new PublicKey(data.subarray(offset, offset + 32));
  offset += 32;

  const bump = data[offset];
  offset += 1;

  // Skip padding: 7 bytes

  return {
    discriminator,
    market,
    mintA,
    mintB,
    lookupTable,
    bump,
  };
}

File: python/src/lightcone_sdk/program/accounts.py
"""Account deserialization for the Lightcone SDK."""

from .constants import (
    EXCHANGE_DISCRIMINATOR,
    EXCHANGE_SIZE,
    GLOBAL_DEPOSIT_TOKEN_DISCRIMINATOR,
    GLOBAL_DEPOSIT_TOKEN_SIZE,
    MARKET_DISCRIMINATOR,
    MARKET_SIZE,
    ORDER_STATUS_DISCRIMINATOR,
    ORDER_STATUS_SIZE,
    ORDERBOOK_DISCRIMINATOR,
    ORDERBOOK_SIZE,
    POSITION_DISCRIMINATOR,
    POSITION_SIZE,
    USER_NONCE_DISCRIMINATOR,
    USER_NONCE_SIZE,
)
from .errors import InvalidAccountDataError, InvalidDiscriminatorError
from .types import (
    Exchange,
    GlobalDepositToken,
    Market,
    MarketStatus,
    Orderbook,
    OrderStatus,
    Position,
    UserNonce,
)
from .utils import decode_bool, decode_pubkey, decode_u8, decode_u16, decode_u64


def _validate_discriminator(data: bytes, expected: bytes, name: str) -> None:
    """Validate account discriminator."""
    if len(data) < 8:
        raise InvalidAccountDataError(f"{name} data too short: {len(data)} bytes")
    actual = data[:8]
    if actual != expected:
        raise InvalidDiscriminatorError(expected, actual)


def deserialize_exchange(data: bytes) -> Exchange:
    """Deserialize an Exchange account.

    Layout (120 bytes):
    - [0..8]: discriminator
    - [8..40]: authority (Pubkey)
    - [40..72]: operator (Pubkey)
    - [72..104]: manager (Pubkey)
    - [104..112]: market_count (u64 LE)
    - [112]: paused (bool)
    - [113]: bump (u8)
    - [114..116]: deposit_token_count (u16 LE)
    - [116..120]: padding
    """
    _validate_discriminator(data, EXCHANGE_DISCRIMINATOR, "Exchange")

    if len(data) < EXCHANGE_SIZE:
        raise InvalidAccountDataError(
            f"Exchange data too short: {len(data)} bytes (expected {EXCHANGE_SIZE})"
        )

    return Exchange(
        authority=decode_pubkey(data, 8),
        operator=decode_pubkey(data, 40),
        manager=decode_pubkey(data, 72),
        market_count=decode_u64(data, 104),
        paused=decode_bool(data, 112),
        bump=decode_u8(data, 113),
        deposit_token_count=decode_u16(data, 114),
    )


def deserialize_market(data: bytes) -> Market:
    """Deserialize a Market account.

    Layout (120 bytes):
    - [0..8]: discriminator
    - [8..16]: market_id (u64 LE)
    - [16]: num_outcomes (u8)
    - [17]: status (u8: 0=Pending, 1=Active, 2=Resolved, 3=Cancelled)
    - [18]: winning_outcome (u8, 255 if not resolved)
    - [19]: has_winning_outcome (bool)
    - [20]: bump (u8)
    - [21..24]: padding (3 bytes)
    - [24..56]: oracle (Pubkey)
    - [56..88]: question_id (32 bytes)
    - [88..120]: condition_id (32 bytes)
    """
    _validate_discriminator(data, MARKET_DISCRIMINATOR, "Market")

    if len(data) < MARKET_SIZE:
        raise InvalidAccountDataError(
            f"Market data too short: {len(data)} bytes (expected {MARKET_SIZE})"
        )

    return Market(
        market_id=decode_u64(data, 8),
        num_outcomes=decode_u8(data, 16),
        status=MarketStatus(decode_u8(data, 17)),
        winning_outcome=decode_u8(data, 18),
        has_winning_outcome=decode_bool(data, 19),
        bump=decode_u8(data, 20),
        oracle=decode_pubkey(data, 24),
        question_id=data[56:88],
        condition_id=data[88:120],
    )


def deserialize_position(data: bytes) -> Position:
    """Deserialize a Position account.

    Layout (80 bytes):
    - [0..8]: discriminator
    - [8..40]: owner (Pubkey)
    - [40..72]: market (Pubkey)
    - [72]: bump (u8)
    - [73..80]: padding (7 bytes)
    """
    _validate_discriminator(data, POSITION_DISCRIMINATOR, "Position")

    if len(data) < POSITION_SIZE:
        raise InvalidAccountDataError(
            f"Position data too short: {len(data)} bytes (expected {POSITION_SIZE})"
        )

    return Position(
        owner=decode_pubkey(data, 8),
        market=decode_pubkey(data, 40),
        bump=decode_u8(data, 72),
    )


def deserialize_order_status(data: bytes) -> OrderStatus:
    """Deserialize an OrderStatus account.

    Layout (32 bytes):
    - [0..8]: discriminator
    - [8..16]: remaining (u64 LE)
    - [16..24]: base_remaining (u64 LE)
    - [24]: is_cancelled (bool)
    - [25..32]: padding (7 bytes)
    """
    _validate_discriminator(data, ORDER_STATUS_DISCRIMINATOR, "OrderStatus")

    if len(data) < ORDER_STATUS_SIZE:
        raise InvalidAccountDataError(
            f"OrderStatus data too short: {len(data)} bytes (expected {ORDER_STATUS_SIZE})"
        )

    return OrderStatus(
        remaining=decode_u64(data, 8),
        base_remaining=decode_u64(data, 16),
        is_cancelled=decode_bool(data, 24),
    )


def deserialize_user_nonce(data: bytes) -> UserNonce:
    """Deserialize a UserNonce account.

    Layout (16 bytes):
    - [0..8]: discriminator
    - [8..16]: nonce (u64 LE)
    """
    _validate_discriminator(data, USER_NONCE_DISCRIMINATOR, "UserNonce")

    if len(data) < USER_NONCE_SIZE:
        raise InvalidAccountDataError(
            f"UserNonce data too short: {len(data)} bytes (expected {USER_NONCE_SIZE})"
        )

    return UserNonce(
        nonce=decode_u64(data, 8),
    )


def deserialize_orderbook(data: bytes) -> Orderbook:
    """Deserialize an Orderbook account.

    Layout (144 bytes):
    - [0..8]: discriminator
    - [8..40]: market (Pubkey)
    - [40..72]: mint_a (Pubkey)
    - [72..104]: mint_b (Pubkey)
    - [104..136]: lookup_table (Pubkey)
    - [136]: base_index (u8)
    - [137]: bump (u8)
    - [138..144]: padding (6 bytes)
    """
    _validate_discriminator(data, ORDERBOOK_DISCRIMINATOR, "Orderbook")

    if len(data) < ORDERBOOK_SIZE:
        raise InvalidAccountDataError(
            f"Orderbook data too short: {len(data)} bytes (expected {ORDERBOOK_SIZE})"
        )

    return Orderbook(
        market=decode_pubkey(data, 8),
        mint_a=decode_pubkey(data, 40),
        mint_b=decode_pubkey(data, 72),
        lookup_table=decode_pubkey(data, 104),
        base_index=decode_u8(data, 136),
        bump=decode_u8(data, 137),
    )


def is_exchange_account(data: bytes) -> bool:
    """Check if data has the Exchange discriminator."""
    return len(data) >= 8 and data[:8] == EXCHANGE_DISCRIMINATOR


def is_market_account(data: bytes) -> bool:
    """Check if data has the Market discriminator."""
    return len(data) >= 8 and data[:8] == MARKET_DISCRIMINATOR


def is_position_account(data: bytes) -> bool:
    """Check if data has the Position discriminator."""
    return len(data) >= 8 and data[:8] == POSITION_DISCRIMINATOR


def is_order_status_account(data: bytes) -> bool:
    """Check if data has the OrderStatus discriminator."""
    return len(data) >= 8 and data[:8] == ORDER_STATUS_DISCRIMINATOR


def is_user_nonce_account(data: bytes) -> bool:
    """Check if data has the UserNonce discriminator."""
    return len(data) >= 8 and data[:8] == USER_NONCE_DISCRIMINATOR


def is_orderbook_account(data: bytes) -> bool:
    """Check if data has the Orderbook discriminator."""
    return len(data) >= 8 and data[:8] == ORDERBOOK_DISCRIMINATOR


def is_global_deposit_token(data: bytes) -> bool:
    """Check if data has the GlobalDepositToken discriminator."""
    return len(data) >= 8 and data[:8] == GLOBAL_DEPOSIT_TOKEN_DISCRIMINATOR


def deserialize_global_deposit_token(data: bytes) -> GlobalDepositToken:
    """Deserialize a GlobalDepositToken account.

    Layout (48 bytes):
    - [0..8]: discriminator
    - [8..40]: mint (Pubkey)
    - [40]: active (bool)
    - [41]: bump (u8)
    - [42..44]: index (u16 LE)
    - [44..48]: padding
    """
    _validate_discriminator(data, GLOBAL_DEPOSIT_TOKEN_DISCRIMINATOR, "GlobalDepositToken")

    if len(data) < GLOBAL_DEPOSIT_TOKEN_SIZE:
        raise InvalidAccountDataError(
            f"GlobalDepositToken data too short: {len(data)} bytes (expected {GLOBAL_DEPOSIT_TOKEN_SIZE})"
        )

    return GlobalDepositToken(
        mint=decode_pubkey(data, 8),
        active=decode_bool(data, 40),
        bump=decode_u8(data, 41),
        index=decode_u16(data, 42),
    )
